oap-server/server-storage-plugin/storage-iotdb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/iotdb/IoTDBClient.java

```diff
@@ -0,0 +1,352 @@
+            healthChecker.unHealth(e);
+            throw new IOException(e);
+        }
+    }
+
+    /**
+     * Normal filter query for a list of data. querySQL must contain "align by device"
+     *
+     * @param modelName      model name
+     * @param querySQL       the SQL for query which must contain "align by device"
+     * @param storageBuilder storage builder for transforming storage result map to entity
+     * @return a list of result data
+     * @throws IOException IoTDBConnectionException or StatementExecutionException
+     */
+    public List<? super StorageData> filterQuery(String modelName, String querySQL,
+                                                 StorageHashMapBuilder<? extends StorageData> storageBuilder)
+            throws IOException {
+        if (!querySQL.contains("align by device")) {
+            throw new IOException("querySQL must contain \"align by device\"");
+        }
+        SessionDataSetWrapper wrapper = null;
+        List<? super StorageData> storageDataList = new ArrayList<>();
+        try {
+            wrapper = sessionPool.executeQueryStatement(querySQL);
+            if (log.isDebugEnabled()) {
+                log.debug("SQL: {}, columnNames: {}", querySQL, wrapper.getColumnNames());
+            }
+
+            List<String> columnNames = wrapper.getColumnNames();
+            IoTDBTableMetaInfo tableMetaInfo = IoTDBTableMetaInfo.get(modelName);
+            List<String> indexes = tableMetaInfo.getIndexes();
+            while (wrapper.hasNext()) {
+                Map<String, Object> map = new ConcurrentHashMap<>();
+                RowRecord rowRecord = wrapper.next();
+                List<Field> fields = rowRecord.getFields();
+                // transform timestamp to time_bucket
+                if (!UITemplate.INDEX_NAME.equals(modelName)) {
+                    map.put(IoTDBClient.TIME_BUCKET, TimeBucket.getTimeBucket(rowRecord.getTimestamp(),
+                            tableMetaInfo.getModel().getDownsampling()));
+                }
+                // field.get(0) -> Device, transform layerName to indexValue
+                String[] layerNames = fields.get(0).getStringValue().split("\\" + IoTDBClient.DOT + "\"");
+                for (int i = 0; i < indexes.size(); i++) {
+                    map.put(indexes.get(i), layerName2IndexValue(layerNames[i + 1]));
+                }
+                for (int i = 0; i < columnNames.size() - 2; i++) {
+                    String columnName = columnNames.get(i + 2);
+                    Field field = fields.get(i + 1);
+                    if (field.getDataType() == null) {
+                        continue;
+                    }
+                    if (field.getDataType().equals(TSDataType.TEXT)) {
+                        map.put(columnName, field.getStringValue());
+                    } else {
+                        map.put(columnName, field.getObjectValue(field.getDataType()));
+                    }
```

Comment:
You may need to check whether the `field` is null or not, it seems `field.getStringValue()` returns a string `"null"` if `field` is null

Reply:
Specifically, this might cause the service group name to be “null” in UI topology.

Reply:
`field` won't be null. What's more, if IoTDB stores a null value, `field.getDataType()` will return null and this condition will be deal with in the line 223.

Reply:
Please check locally. The group in topology is "null", which is wrong

Reply:
I didn't dig deeper to see what's the problem but "null" string exists indeed

Reply:
Or maybe there are some places where you use `String.valueOf()` without checking nullability

Reply:
When inserting data, the `service_group` will be "null" if user doesn't set since `service_group` is one of the index and its null value is casted to "null". I have located this bug and will fix it.